### PR TITLE
persisted storage transfer page hashmap

### PR DESCRIPTION
**Pull request by Wenbin1002**
## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [ ] Improvement
- [ ] Documentation
- [x] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #16953

## What this PR does / why we need it:

减少transfer page hashmap的内存消耗

在 sysbench 100w 1000 threads bulk insert 测试场景下，transfer page内存消耗从40g降低至4g
企业微信截图_17206645205114
企业微信截图_1720666501392

grafana
image

___

### **PR Type**
Feature, Enhancement, Tests, Configuration changes

___

### **Description**
- Introduced `HashPageMap` type with protobuf support.
- Enhanced `TransferHashPage` with atomic hashmap and file service.
- Integrated transfer page management with file service.
- Refactored `TransferTable` and updated TTL handling.
- Added and updated transfer page metrics in Grafana.
- Added tests for transfer table persistence and clearing.
- Updated transfer tests with new Train method.
- Added command for setting transfer TTL in inspect tool.
- Handled transfer pages in flush table tail.
- Initialized local file service and updated transfer table.
- Added `LocalFs` field to `Options` struct.
- Set default value for `TransferInterval`.
- Registered new transfer page metrics.
- Added `HashPageMap` message definition.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Feature</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>api.pb.go</strong><dd><code>Introduce `HashPageMap` type with protobuf support.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pkg/pb/api/api.pb.go

<li>Added a new <code>HashPageMap</code> type with associated methods for marshaling, <br>unmarshaling, and protobuf operations.<br> <li> Registered the new <code>HashPageMap</code> type and its map type in the protobuf <br>initialization.<br> <li> Implemented methods for <code>HashPageMap</code> to handle protobuf serialization <br>and deserialization.<br>

</details>

  </td>
  <td><a href="https://github.com/matrixorigin/matrixone/pull/17490/files#diff-1bb52ddffebfd489bd487114477c966c402ce9c4aebc82385b23f1ac6f58510e">+432/-146</a></td>

</tr>                    
</table></details></td></tr><tr><td><strong>Enhancement</strong></td><td><details><summary>12 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>pages.go</strong><dd><code>Enhance `TransferHashPage` with atomic hashmap and file service.</code></dd></summary>
<hr>

pkg/vm/engine/tae/model/pages.go

<li>Introduced a new <code>TransferHashPage</code> structure with atomic pointer for <br><code>hashmap</code>.<br> <li> Added methods for managing the lifecycle and state of <br><code>TransferHashPage</code>, including TTL, clearing, and persistence.<br> <li> Implemented file service integration for loading and saving <br><code>TransferHashPage</code> data.<br>

</details>

  </td>
  <td><a href="https://github.com/matrixorigin/matrixone/pull/17490/files#diff-44086321e3493435f1083550a3a7c8d9f6a2799300ff20ac03d3bb41ef08c4fb">+175/-12</a></td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>mergeobjects.go</strong><dd><code>Integrate transfer page management with file service.</code>&nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pkg/vm/engine/tae/tables/txnentries/mergeobjects.go

<li>Added functions to initialize, add, and write transfer pages using <br>file service.<br> <li> Enhanced <code>prepareTransferPage</code> method to handle new <code>TransferHashPage</code> <br>structure and file service operations.<br>

</details>

  </td>
  <td><a href="https://github.com/matrixorigin/matrixone/pull/17490/files#diff-ef36687940ace38467e7048dd18a1cbc36d8eeeefc02ce86fd606f7d0cd30ee6">+71/-2</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>transfer.go</strong><dd><code>Refactor TransferTable and update TTL handling</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pkg/vm/engine/tae/model/transfer.go

<li>Refactored <code>TransferTable</code> to remove TTL parameter and manage deleted <br>pages.<br> <li> Modified <code>prepareTTL</code> and <code>executeTTL</code> methods to handle memory and disk <br>separately.<br> <li> Updated metrics and added new ones for transfer table operations.<br>

</details>

  </td>
  <td><a href="https://github.com/matrixorigin/matrixone/pull/17490/files#diff-c115046f558a4d4367cfe9d23e07b2ace0ff013ac60616565975d94f5ce83ffd">+38/-31</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>grafana_dashboard_task.go</strong><dd><code>Add and update transfer page metrics in Grafana</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pkg/util/metric/v2/dashboard/grafana_dashboard_task.go

<li>Added new metrics for transfer page operations in Grafana dashboard.<br> <li> Updated existing metric names and descriptions.<br>

</details>

  </td>
  <td><a href="https://github.com/matrixorigin/matrixone/pull/17490/files#diff-83c0737cffe25114938fc3ffa20d611230de6fc7447d045a176409ecb53c8a9f">+66/-4</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>inspect.go</strong><dd><code>Add command for setting transfer TTL in inspect tool</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pkg/vm/engine/tae/rpc/inspect.go

<li>Added new <code>transferArg</code> struct and related command for setting transfer <br>TTL.<br> <li> Integrated new command into the inspection context.<br>

</details>

  </td>
  <td><a href="https://github.com/matrixorigin/matrixone/pull/17490/files#diff-a7350916eefe8e5ceea15f65874fa0f87b7e68c2127ea19dd0ce33b5520ccf5f">+44/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>flushTableTail.go</strong><dd><code>Handle transfer pages in flush table tail</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pkg/vm/engine/tae/tables/txnentries/flushTableTail.go

<li>Added logic to handle transfer pages during table tail flush.<br> <li> Introduced latency measurement for transfer page operations.<br>

</details>

  </td>
  <td><a href="https://github.com/matrixorigin/matrixone/pull/17490/files#diff-332ca4c3c8fedfcc6098138df055ae1ad545222f5f00c7c2a60f1d41e7df8b22">+26/-2</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>task.go</strong><dd><code>Add new metrics for transfer page operations</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pkg/util/metric/v2/task.go

<li>Added new histograms for transfer page latencies and hit counts.<br> <li> Renamed existing gauge for transfer page size.<br>

</details>

  </td>
  <td><a href="https://github.com/matrixorigin/matrixone/pull/17490/files#diff-3100c6cb139a6482306d0a557727df768e3983c5e6ab545fc9582c61cd5cf828">+31/-1</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>open.go</strong><dd><code>Initialize local file service and update transfer table</code>&nbsp; &nbsp; </dd></summary>
<hr>

pkg/vm/engine/tae/db/open.go

<li>Added initialization for local file service.<br> <li> Updated transfer table creation and cron job for TTL handling.<br>

</details>

  </td>
  <td><a href="https://github.com/matrixorigin/matrixone/pull/17490/files#diff-70f750e6241afa9a68b84ae68293c78c247640ee87b5225c5d3b7265ca531797">+11/-3</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>factory.go</strong><dd><code>Initialize local file service in TAE storage</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pkg/tnservice/factory.go

- Added local file service initialization in TAE storage creation.

</details>

  </td>
  <td><a href="https://github.com/matrixorigin/matrixone/pull/17490/files#diff-e7d2f3a8f4512a258961f6dd069947e6580dbee38232b71ba792966511254477">+7/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>types.go</strong><dd><code>Add LocalFs field to Options struct</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pkg/vm/engine/tae/options/types.go

- Added `LocalFs` field to `Options` struct.

</details>

  </td>
  <td><a href="https://github.com/matrixorigin/matrixone/pull/17490/files#diff-1c42bccfaae4daf20288fbaf1a80905d600284be96bec190643fd6496eae5bf2">+2/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>metrics.go</strong><dd><code>Register new transfer page metrics</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pkg/util/metric/v2/metrics.go

- Registered new transfer page metrics.

</details>

  </td>
  <td><a href="https://github.com/matrixorigin/matrixone/pull/17490/files#diff-97ea305cc79979fa7db0a8fd618613cdda1c9bda32e6aa91847d8a754de671f3">+3/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>api.proto</strong><dd><code>Add HashPageMap message definition</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

proto/api.proto

- Added new `HashPageMap` message definition.

</details>

  </td>
  <td><a href="https://github.com/matrixorigin/matrixone/pull/17490/files#diff-ecf0878b123623e280a40f3e3d75af4267f11cfe7138753e671654c0ba05f228">+4/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></details></td></tr><tr><td><strong>Tests</strong></td><td><details><summary>2 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>db_test.go</strong><dd><code>Add tests for transfer table persistence and clearing</code>&nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pkg/vm/engine/tae/db/test/db_test.go

<li>Added new imports for <code>errors</code> and <code>model</code>.<br> <li> Introduced two new test functions: <code>TestPersistTransferTable</code> and <br><code>TestClearPersistTransferTable</code>.<br> <li> Implemented logic for creating and managing transfer pages in tests.<br>

</details>

  </td>
  <td><a href="https://github.com/matrixorigin/matrixone/pull/17490/files#diff-858060e31b5a9bf1276488170b96f1f70e3d50b9ef08c44bba5d7acf5778d0a3">+129/-0</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>transfer_test.go</strong><dd><code>Update transfer tests with new Train method</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pkg/vm/engine/tae/model/transfer_test.go

<li>Updated tests to use new <code>Train</code> method with map input.<br> <li> Adjusted TTL checks and added latency measurements.<br>

</details>

  </td>
  <td><a href="https://github.com/matrixorigin/matrixone/pull/17490/files#diff-e3022daacb0913df74d0f17f7091f86f5df3d3d87492bfa8d687589996adf725">+17/-10</a>&nbsp; </td>

</tr>                    
</table></details></td></tr><tr><td><strong>Configuration changes</strong></td><td><details><summary>2 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>options.go</strong><dd><code>Set default value for TransferInterval</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pkg/vm/engine/tae/options/options.go

- Added default value for `TransferInterval` in options.

</details>

  </td>
  <td><a href="https://github.com/matrixorigin/matrixone/pull/17490/files#diff-7a129eb095032535e9454cee079a7ead8816ab669e0ee616b78e020cb6191364">+3/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>cfg.go</strong><dd><code>Add TransferInterval to CheckpointCfg</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pkg/vm/engine/tae/options/cfg.go

- Added `TransferInterval` field to `CheckpointCfg` struct.

</details>

  </td>
  <td><a href="https://github.com/matrixorigin/matrixone/pull/17490/files#diff-3c4d928886e9865142ec9e96b137c1f7e34ad742d220d580b0f1dc6db6c0f208">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></details></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

**Review by aptend — APPROVED** [review]

File: pkg/vm/engine/tae/tables/txnentries/mergeobjects.go [hunk 1]
```diff
@@ -371,3 +398,45 @@ func (entry *mergeObjectsEntry) PrepareCommit() (err error) {
...
+		Offset: offset,
+		Size:   int64(len(data)),
+		Data:   data,
+	}
+	ioVector.Entries = append(ioVector.Entries, ioEntry)
+
+	return nil
+}
+
+func WriteTransferPage(pages []*model.TransferHashPage, ioVector fileservice.IOVector) error {
```
Comment (XuPeng-SH) [inline pkg/vm/engine/tae/tables/txnentries/mergeobjects.go:427]: use `ctx context.Context` as the first input parameter

File: pkg/vm/engine/tae/tables/txnentries/mergeobjects.go [hunk 2]
```diff
@@ -95,6 +98,10 @@ func NewMergeObjectsEntry(
 func (entry *mergeObjectsEntry) prepareTransferPage() {
```
Comment (XuPeng-SH) [inline pkg/vm/engine/tae/tables/txnentries/mergeobjects.go:98]: add `ctx context.Context`. we need to pass the `ctx` from outside to the inner.

File: pkg/vm/engine/tae/tables/txnentries/mergeobjects.go [hunk 3]
```diff
@@ -106,14 +113,34 @@ func (entry *mergeObjectsEntry) prepareTransferPage() {
...
+		start = time.Now()
+		err := WriteTransferPage(pages, *ioVector)
+		if err != nil {
+			return
+		}
+		duration += time.Since(start)
+		v2.TransferPageMergeLatencyHistogram.Observe(duration.Seconds())
 	}
 	if k != len(entry.transMappings.Mappings) {
 		panic(fmt.Sprintf("k %v, mapping %v", k, len(entry.transMappings.Mappings)))
```
Comment (XuPeng-SH) [inline pkg/vm/engine/tae/tables/txnentries/mergeobjects.go:146]: if it is unexpected, we can fatal this message and crash. For the background task, we never panic but crash the process.

File: pkg/vm/engine/tae/tables/txnentries/mergeobjects.go (hunk 2 above)
Comment (XuPeng-SH) [inline pkg/vm/engine/tae/tables/txnentries/mergeobjects.go:98]: I think it should return `err error` now and the caller need to handle error scenario

File: pkg/vm/engine/tae/tables/txnentries/mergeobjects.go [hunk 4]
```diff
@@ -106,14 +113,34 @@ func (entry *mergeObjectsEntry) prepareTransferPage() {
...
+
 			entry.pageIds = append(entry.pageIds, id)
 			_ = entry.rt.TransferTable.AddPage(page)
+			pages = append(pages, page)
 		}
+
+		start = time.Now()
+		err := WriteTransferPage(pages, *ioVector)
+		if err != nil {
+			return
```
Comment (XuPeng-SH) [inline pkg/vm/engine/tae/tables/txnentries/mergeobjects.go:140]: Any error, return ok? i think we need to inject some error to see whether we handle error as expected. (memory or storage leak, panic, or something horrible)

File: pkg/vm/engine/tae/db/open.go [hunk 5]
```diff
@@ -93,6 +93,14 @@ func Open(ctx context.Context, dirname string, opts *options.Options) (db *DB, e
 		// TODO:fileservice needs to be passed in as a parameter
 		opts.Fs = objectio.TmpNewFileservice(ctx, path.Join(dirname, "data"))
 	}
+	if opts.LocalFs == nil {
+		opts.LocalFs = objectio.TmpNewFileservice(ctx, path.Join(dirname, "data"))
+	}
+	model.SetFileService(opts.LocalFs)
```
Comment (XuPeng-SH) [inline pkg/vm/engine/tae/db/open.go:99]: don't introduce any global variable any more

**Review by XuPeng-SH — CHANGES_REQUESTED** [review]
please modify as commented

**Review by XuPeng-SH — APPROVED** [review]

**Review by zhangxu19830126 — APPROVED** [review]

**Pull request merged**